File: algorithms/3_hard/M/01289_minimum_falling_path_sum_ii/dp.c
struct tuple {
    int min1;
    int min2;
    int min1_idx;
};

inline static struct tuple calc_min1_min2(const int *restrict arr, const int len) {
    int min1 = INT_MAX;
    int min2 = INT_MAX;
    int min1_idx = 0;

    for (int i = 0; i < len; i += 1) {
        const int v = arr[i];
        if (v <= min1) {
            min2 = min1;
            min1_idx = i;
            min1 = v;
        } else if (v <= min2) {
            min2 = v;
        }
    }

    return (struct tuple){ .min1 = min1, .min2 = min2, .min1_idx = min1_idx };
}

int minFallingPathSum(int *restrict *restrict grid, const int rows, const int *restrict cols_data) {
    assert(rows > 0 && rows == cols_data[0]);

    const int n = rows;
    if (n == 1) { return grid[0][0]; }

    int *restrict buff = calloc((size_t)n * 2, sizeof(*buff));
    assert(buff != NULL);
    int *restrict curr = &buff[0];
    int *restrict prev = &buff[n];

    for (int i = n - 1; i >= 0; i -= 1) {
        const int *restrict g = grid[i];
        const struct tuple data_mins = calc_min1_min2(prev, n);
        for (int j = 0; j < n; j += 1) {
            curr[j] = g[j] + data_mins.min1;
        }
        curr[data_mins.min1_idx] = g[data_mins.min1_idx] + data_mins.min2;

        void *tmp = curr;
        curr = prev;
        prev = tmp;
    }

    int min = INT_MAX;
    for (int i = 0; i < n; i += 1) {
        min = MIN(min, prev[i]);
    }

    free(buff);
    return min;
}

File: algorithms/3_hard/M/01289_minimum_falling_path_sum_ii/dp.rs
impl Solution {
    #[inline(always)]
    fn calc_min1_min2(arr: &[i32]) -> (i32, i32, usize) {
        let (mut m1, mut m2) = (i32::MAX, i32::MAX);
        let mut m1_idx = 0_usize;

        for (i, &v) in arr.iter().enumerate() {
            if v <= m1 {
                m2 = m1;
                m1_idx = i;
                m1 = v;
            } else if v <= m2 {
                m2 = v;
            }
        }

        (m1, m2, m1_idx)
    }

    pub fn min_falling_path_sum(grid: Vec<Vec<i32>>) -> i32 {
        assert!(grid.len() > 0 && grid.len() == grid[0].len());

        let n = grid[0].len();
        if n == 1 { return grid[0][0]; }
        let mut buff = vec![0_i32; 2 * n];
        let (mut curr, mut prev) = buff.split_at_mut(n);

        for g in grid.into_iter().rev() {
            let (n1, n2, n1_idx) = Self::calc_min1_min2(prev);
            for i in 0..n {
                curr[i] = g[i] + n1;
            }
            curr[n1_idx] = g[n1_idx] + n2;

            (prev, curr) = (curr, prev);
        }

        *prev.into_iter().min().unwrap()
    }
}
